src/exporters/xray/transformer.rs

```diff
@@ -33,47 +46,72 @@ impl TransformPayload<ResourceSpans> for Transformer {
-
-        for message in messages {
-            // Extract metadata from message
-            if let Some(metadata) = message.metadata {
-                all_metadata.push(metadata);
-            }
+    ) -> Result<Vec<XRayValuePayload>, ExportError> {
+        let mut all_payloads = Vec::new();
+        const MAX_SPANS_PER_CHUNK: usize = 50;
+
+        // Buffer to accumulate spans across messages
+        let mut span_buffer = Vec::new();
+        // Track metadata from messages that contributed spans to the current buffer
+        let mut metadata_tracking: Vec<MessageMetadata> = Vec::new();
 
-            // Process the spans
+        for mut message in messages.into_iter() {
+            // Transform all spans from this message
             for rs in message.payload {
                 for ss in rs.scope_spans {
                     for span in ss.spans {
                         match self.transformer.apply(span) {
-                            Ok(v) => payload.push(v),
-                            Err(e) => {
-                                return (
-                                    Err(e),
-                                    if all_metadata.is_empty() {
+                            Ok(v) => {
+                                span_buffer.push(v);
+
+                                // If buffer is full, create a request
+                                if span_buffer.len() == MAX_SPANS_PER_CHUNK {
+                                    let chunk_value =
+                                        Value::Array(std::mem::take(&mut span_buffer));
+
+                                    // Record this message's contribution if it has metadata
+                                    if let Some(metadata) = message.metadata.as_ref() {
+                                        metadata_tracking.push(metadata.clone());
+                                    }
+
+                                    // Build metadata list from all contributing messages
+                                    let chunk_metadata = if metadata_tracking.is_empty() {
                                         None
                                     } else {
-                                        Some(all_metadata)
-                                    },
-                                );
+                                        Some(std::mem::take(&mut metadata_tracking))
+                                    };
+
+                                    all_payloads
+                                        .push(XRayValuePayload::new(chunk_value, chunk_metadata));
+                                }
                             }
+                            Err(e) => return Err(e),
                         }
                     }
                 }
             }
+
+            // Track this message's metadata for the current buffer if it has metadata
+            if let Some(metadata) = message.metadata.take() {
+                metadata_tracking.push(metadata);
+            }
```

Comment:
Could this get here and span_buffer be empty if there were exactly N * 50 spans contained in the message? This would potentially add the current message's metadata onto the next message's span_buffer, even though there are no spans from it in the remaining buffer?

Reply:
Good catch, added a unit test for this one.